src/Microsoft.ML.TimeSeries/STL/InnerStl.cs

```diff
@@ -0,0 +1,351 @@
+        {
+            for (int i = 0; i < _length; i++)
+            {
+                s[i] = c[i] - lowPass.Y[i];
+            }
+        }
+
+        private void Deseasonalizing(IReadOnlyList<double> y, double[] s, double[] deseasonSeries)
+        {
+            for (int i = 0; i < _length; i++)
+            {
+                deseasonSeries[i] = y[i] - s[i];
+            }
+        }
+
+        private void TrendSmooth(double[] deseasonSeries, int np, double[] t)
+        {
+            List<double> virtualDeseasonSeries = VirtualXValuesProvider.GetXValues(deseasonSeries.Length);
+            FastLoess trender = new FastLoess(virtualDeseasonSeries, deseasonSeries, _isTemporal, Nt(np));
+            trender.Estimate();
+            for (int i = 0; i < _length; i++)
+            {
+                t[i] = trender.Y[i];
+            }
+        }
+
+        /// <summary>
+        /// this class provides the virtual x values for multi object usage.
+        /// the cache mechanism is used for performance consideration.
+        /// </summary>
+        internal class VirtualXValuesProvider
+        {
+            private static Dictionary<int, List<double>> _xValuesPool;
+
+            static VirtualXValuesProvider()
+            {
+                _xValuesPool = new Dictionary<int, List<double>>();
+            }
+
+            /// <summary>
+            /// get a list of virtual x-axis values. the values are from 0 to length - 1.
+            /// </summary>
+            /// <param name="length">specify the length you want to create the x values.</param>
+            /// <returns>if this is cached, return directly. otherwise, create a new list and return</returns>
+            internal static List<double> GetXValues(int length)
+            {
+                lock (_xValuesPool)
+                {
+                    List<double> xValues;
+                    if (_xValuesPool.TryGetValue(length, out xValues))
+                        return xValues;
+
+                    var newXValues = new List<double>(length);
+                    for (int i = 0; i < length; i++)
+                        newXValues.Add(i);
+
+                    _xValuesPool.Add(length, newXValues);
+                    return newXValues;
+                }
+            }
+        }
+
+        private static List<double> MovingAverage(IReadOnlyList<double> s, int length)
+        {
+            List<double> results = new List<double>(s.Count);
+            double partialSum = 0;
+            for (int i = 0; i < length; ++i)
+            {
+                partialSum += s[i];
+            }
+
+            for (int i = length; i < s.Count; ++i)
+            {
+                results.Add(partialSum / length);
+                partialSum = partialSum - s[i - length] + s[i];
+            }
+            results.Add(partialSum / length);
+
+            return results;
+        }
+
+        private void AllocateDoubleArray(ref double[] arr, int length)
```

Comment:
The same comment as in:
https://github.com/dotnet/machinelearning/pull/5202/files?file-filters%5B%5D=.cs&file-filters%5B%5D=.csv#r440452024